Users with no self-service access may update pivot tables and leave them in a non-working state

### Describe the bug

Users with no self-service access are allowed to update visualization settings of pivot tables, and they may save cards in a non-workable state with corrupted visualization settings.

### To Reproduce

1. Create a pivot table, on the sample database: Products, Count, Grouped by Category and Rating. Save it.
2. Open it with another user with no self-service access to Products
3. Open visualization settings, and rearrange Rows / Columns / Measures
4. (!) We get the error `You do not have permissions to run this query`.
<img width="684" alt="image" src="https://github.com/metabase/metabase/assets/11080638/c1ed5b75-f050-4451-9305-d04df5becbca">

5. The question becomes dirty, so try to Save it and a Save new question dialog box appears.
 Hit Save again, the original question will be overwritten.  The incorrect dialog error in this step is reported separately in https://github.com/metabase/metabase/issues/37373)

<img width="652" alt="image" src="https://github.com/metabase/metabase/assets/11080638/83c8a574-c645-467b-8f43-b758745e7abd"> 

6. (!) The card is now broken, with incomplete visualization settings.
<img width="150" alt="image" src="https://github.com/metabase/metabase/assets/11080638/108249ef-9163-4859-859f-2a3c620a946e">

With the no self-service access user it is not possible to return the card into a working state.

### Expected behavior

_No response_

### Logs

Console log:
```
session-middleware.js:53 Error: Invalid metadata returned for cell 1:
        x:104, y:0, width:NaN, height:30
    at t.cellCount (calculateSizeAndPositionData.js:17:13)
    at t.value (calculateSizeAndPositionData.js:35:1)
    at t.value (CollectionView.js:315:27)
```

api/card response:
```
[...]
    "visualization_settings": {
        "pivot_table.column_split": {
            "rows": [
                [
                    "field",
                    740,
                    {
                        "base-type": "type/Text"
                    }
                ],
                [
                    "field",
                    742,
                    {
                        "base-type": "type/Float",
                        "binning": {
                            "strategy": "default"
                        }
                    }
                ]
            ],
            "columns": [],
            "values": []
        },
[...]
```

### Information about your Metabase installation

```JSON
1.48.1
```


### Severity

A similar issue is blocking some users.

### Additional context

_No response_

--- a/e2e/test/scenarios/visualizations-tabular/pivot_tables.cy.spec.js
+++ b/e2e/test/scenarios/visualizations-tabular/pivot_tables.cy.spec.js
@@ -1,4 +1,4 @@
-import { SAMPLE_DB_ID } from "e2e/support/cypress_data";
+import { SAMPLE_DB_ID, USER_GROUPS } from "e2e/support/cypress_data";
 import { SAMPLE_DATABASE } from "e2e/support/cypress_sample_database";
 import {
   restore,
@@ -15,6 +15,7 @@ import {
   openPublicLinkPopoverFromMenu,
   openStaticEmbeddingModal,
   modal,
+  createQuestion,
 } from "e2e/support/helpers";
 
 const {
@@ -1140,6 +1141,74 @@ describe("scenarios > visualizations > pivot tables", { tags: "@slow" }, () => {
       );
     },
   );
+
+  describe("issue 37380", () => {
+    beforeEach(() => {
+      const categoryField = [
+        "field",
+        PRODUCTS.CATEGORY,
+        { "base-type": "type/Text" },
+      ];
+
+      const createdAtField = [
+        "field",
+        PRODUCTS.CREATED_AT,
+        {
+          "base-type": "type/DateTime",
+          "temporal-unit": "month",
+        },
+      ];
+
+      // to reproduce metabase#37380 it's important that user has access to the database, but not to the table
+      cy.updatePermissionsGraph({
+        [USER_GROUPS.DATA_GROUP]: {
+          [SAMPLE_DB_ID]: {
+            "create-queries": {
+              PUBLIC: {
+                [PRODUCTS_ID]: "no",
+              },
+            },
+          },
+        },
+      });
+
+      createQuestion(
+        {
+          query: {
+            "source-table": PRODUCTS_ID,
+            aggregation: ["count"],
+            breakout: [categoryField, createdAtField],
+          },
+          display: "pivot",
+          visualization_settings: {
+            "pivot_table.column_split": {
+              rows: [createdAtField],
+              columns: [categoryField],
+              values: [["aggregation", 0]],
+            },
+            "pivot_table.column_widths": {
+              leftHeaderWidths: [141],
+              totalLeftHeaderWidths: 141,
+              valueHeaderWidths: {},
+            },
+          },
+        },
+        {
+          wrapId: true,
+          idAlias: "questionId",
+        },
+      );
+    });
+
+    it("does not allow users with no table access to update pivot questions (metabase#37380)", () => {
+      cy.signInAsNormalUser();
+      visitQuestion("@questionId");
+      cy.findByTestId("viz-settings-button").click();
+      cy.findByLabelText("Show row totals").click();
+
+      cy.button("Save").should("have.attr", "disabled");
+    });
+  });
 });
 
 const testQuery = {

--- a/frontend/src/metabase/visualizations/visualizations/PivotTable/settings.ts
+++ b/frontend/src/metabase/visualizations/visualizations/PivotTable/settings.ts
@@ -195,10 +195,14 @@ export const settings = {
         return hasOnlyFormattableColumns && !columnFormat.highlight_row;
       });
     },
-    getProps: (series: Series) => ({
-      canHighlightRow: false,
-      cols: series[0].data?.cols.filter(isFormattablePivotColumn),
-    }),
+    getProps: (series: Series) => {
+      const cols = series[0].data?.cols ?? [];
+
+      return {
+        canHighlightRow: false,
+        cols: cols.filter(isFormattablePivotColumn),
+      };
+    },
     getHidden: ([{ data }]: [{ data: DatasetData }]) =>
       !data?.cols.some(col => isFormattablePivotColumn(col)),
   },

--- a/frontend/src/metabase/visualizations/visualizations/PivotTable/settings.ts
+++ b/frontend/src/metabase/visualizations/visualizations/PivotTable/settings.ts
@@ -197,9 +197,7 @@ export const settings = {
     },
     getProps: (series: Series) => ({
       canHighlightRow: false,
-      cols: (series[0].data.cols as DatasetColumn[]).filter(
-        isFormattablePivotColumn,
-      ),
+      cols: series[0].data?.cols.filter(isFormattablePivotColumn),
     }),
     getHidden: ([{ data }]: [{ data: DatasetData }]) =>
       !data?.cols.some(col => isFormattablePivotColumn(col)),

